tf2onnx/rewriter/random_uniform.py

```diff
@@ -59,25 +60,50 @@ def rewrite_random_uniform_fold_const(g, ops):
-        new_node = create_onnx_random_uniform_op(g, tmax, tmin, ru_op, output)
+        to_delete = list(set(match.get_nodes()))
+        new_node = create_onnx_random_uniform_op(g, tmax, tmin, ru_op, output, to_delete)
         g.replace_all_inputs(ops, output.output[0], new_node.output[0])
-        for n in set(match.get_nodes()):
+        for n in to_delete:
             g.remove_node(n.name)
 
     return ops
 
 
-def create_onnx_random_uniform_op(g, tmax, tmin, ru_op, output):
+def create_onnx_random_uniform_op(g, tmax, tmin, ru_op, output, to_delete):
     dtype = g.get_dtype(output.output[0])
     op_name = utils.make_name("RandomUniform")
-    if ru_op.inputs[0].type == "Shape":
-        shape_node = ru_op.inputs[0]
-        new_node = g.make_node("RandomUniformLike", inputs=[shape_node.input[0]], name=op_name,
-                               attr={"low": tmin, "high": tmax, "dtype": dtype},
-                               shapes=shape_node.output_shapes, dtypes=[dtype])
-    else:
-        shape = g.get_shape(output.output[0])
+    shape_node = ru_op.inputs[0]
+    shape = g.get_shape(output.output[0])
+    if shape_node.is_const():
+        # if the tensorflow input (aka the shape) is const we can use the RandomUniform op
         new_node = g.make_node("RandomUniform", [], name=op_name,
                                attr={"low": tmin, "high": tmax, "dtype": dtype, "shape": shape},
                                shapes=[shape], dtypes=[dtype])
+    else:
+        if shape_node.type == "Shape":
+            # if shape is dynamic - in tensorflow shape comes as tensor VALUE,
+            # in onnx RandomUniformLike finds takes the shape from the tensor itself.
+            # In many cases there is a shape op in tensorflow before RandomUniform and
+            # to make that work for onnx we just need to remove the shape op.
+            new_node = g.make_node("RandomUniformLike", inputs=[shape_node.input[0]], name=op_name,
+                                   attr={"low": tmin, "high": tmax, "dtype": dtype},
+                                   shapes=shape, dtypes=[dtype])
+        else:
+            # if the shape is calculated we need to create a tensor so RandomUniformLike
+            # can take the shape from there. Pre opset9 this is somewhat hacky because there is
+            # no real fill op in onnx. In general this is not going to help performance but the tensors
+            # created are expected to be small.
+
+            # tell the caller to not delete the shape node
+            to_delete.remove(shape_node)
+            # create a fill op with the shape of the value of the input tensor
+            zero = g.make_const(utils.make_name("zero"), np.zeros((), dtype=np.float32))
+            fill_node = g.make_node("Fill", inputs=[shape_node.output[0], zero.name],
+                                    shapes=shape, dtypes=[dtype])
+            func, _ = handler.tf_op.find_effective_op("Fill")

```

Comment:
don't feel quite right about this.
what we want to do is **create an onnx node**, but we have to **create a tf node first and then convert it**...
and what if there is no corresponding "Fill" op in tf?

Reply:
i think we do need an abstraction of onnx node(s) builder/creator.
so we can reuse common onnx node(s) creation routines among mapping functions and handle different opset versions properly inside builder/creator.

we will demonstrate an idea in Slice-10 PR later.

Reply:
its a little wired because instead of composing onnx it composes tf ops which has the side effect that one doesn't need to worry about different opset versions. At least for that part this works good. I should have checked error but thought its given that it is there.
But yes, a more formal pattern would be good so I'll watch slice-10.